keps/sig-node/4369-allow-special-characters-environment-variable/README.md

```diff
@@ -0,0 +1,290 @@
+
+### Version Skew Strategy
+
+kube-apiserver will need to enable feature gates to use this feature.
+
+If kube-apiserver is not enabled feature gate will use strict validation.
+
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RelaxedValidationEnvironment
+  - Components depending on the feature gate: kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+If close the feature gate, already running workloads will not be affected in any way, 
+but cannot create workloads that use special characters as environment variables.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The feature should continue to work just fine.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+When a feature gate is closed, already running workloads are not affected in any way, but update fields for workload will cause the workload to fail.
+
+###### What specific metrics should inform a rollback?
+
+N/A
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+- N/A
+
+### Dependencies
+
+N/A
+
+###### Does this feature depend on any specific services running in the cluster?
+
+No.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+No
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+\- 2023-12-21: Initial draft KEP
+
+## Drawbacks
+
+## Alternatives
+
```

Comment:
My alternatives:
- do nothing (leave it as-is)
- relax the rule, but with a long beta period where the existing rule remains the default.
  Ensure that the beta period doesn't end until ValidatingAdmissionPolicy is GA and has been for 2 minor releases.
  _Clearly_ document how to use a ValidatingAdmissionPolicy to get behavior equivelent to the legacy checking,
  and signpost people to these docs when graduating the looser validation to be the Kubernetes default.
- define a label or annotation for each namespace that controls how Pod environment variables are validated in that namespace
- [more complex!]
  add an API kind to specify the validation rules for Pods

  Create a new API kind, eg PodValidationRule. It's **namespaced**. Within the `.spec` of each object, define:
  - a Pod selector
  - an optional CEL validation rule for environment variable keys
  - an optional CEL validation rule for environment variable values

  If any of the selected validation rules don't pass for a Pod, reject it at admission time. Set up a defaulting
  mechanism to 
  Also, define how Pod templates interact with this new API (eg: you get a `Warning:` when you create
  a Deployment where the PodTemplate inside the Deployment wouldn't pass validation)

However we do this, I like the idea of letting the change be rolled out namespace by namespace. The ValidatingAdmissionPolicy + docs option would allow this.

Reply:
Thank you very much for your additional thoughts.

Reply:
> * relax the rule, but with a long beta period where the existing rule remains the default.

I think this overlaps with what our current KEP is trying to achieve.

Reply:
I'm less scared of this change than @sftim seems to be :)

We could show examples of gatekeeper & kyverno rules which can restrict it back to the old syntax, but I don't know that we need to wait for ValidatingAdmissionPolicy+++ ?

Reply:
These are the alternatives we'd rule out, right? Let's document that we considered them but they aren't our recommendation.

Reply:
ACK